Content.Shared/_Impstation/Replicator/SharedReplicatorNestSystem.cs

```diff
@@ -0,0 +1,336 @@
+        if (ent.Comp.SpawningProgress >= ent.Comp.NextSpawnAt)
+        {
+            SpawnNew(ent);
+            ent.Comp.NextSpawnAt += ent.Comp.SpawnNewAt;
+        }
+
+        // and dirty so the client knows if it's supposed to update the nest visuals
+        Dirty(ent);
+    }
+
+    private void SpawnNew(Entity<ReplicatorNestComponent> ent)
+    {
+        // SUPER don't run this clientside
+        if (_net.IsClient)
+            return;
+
+        // spawn a new replicator
+        var spawner = Spawn(ent.Comp.ToSpawn, Transform(ent).Coordinates);
+        // TODO:
+        //OnSpawnTile(ent, ent.comp.Level * 2, "FloorReplicator");
+
+        // make sure our new GhostRoleSpawnPoint knows where it came from, so it can pass that down to the replicator it spawns.
+        var tracker = EnsureComp<SpawnedFromTrackerComponent>(spawner);
+        tracker.SpawnedFrom = ent;
+
+        ent.Comp.UnclaimedSpawners.Add(spawner);
+    }
+
+    public void UpgradeAll(Entity<ReplicatorNestComponent> ent)
+    {
+        // don't run this clientside
+        if (_net.IsClient || !_timing.IsFirstTimePredicted)
+            return;
+
+        foreach (var replicator in ent.Comp.SpawnedMinions)
+        {
+            if (!TryComp<ReplicatorComponent>(replicator, out var comp) || comp == null)
+                continue;
+
+            if (comp.UpgradeStage >= MaxUpgradeStage || comp.TargetUpgradeStage >= MaxUpgradeStage)
+                continue;
+
+            if (!TryComp<MindContainerComponent>(replicator, out var mindContainer) || mindContainer.Mind == null)
+                continue;
+
+            comp.TargetUpgradeStage++;
+
+            var targetAction = comp.TargetUpgradeStage == 1 ? comp.Level2Action : comp.Level3Action;
+
+            if (!mindContainer.HasMind)
+                _actions.AddAction(replicator, targetAction);
+            else if (mindContainer.Mind != null)
+                _actionContainer.AddAction((EntityUid)mindContainer.Mind, targetAction);
+        }
+    }
+
+    public void OnUpgrade2(Entity<ReplicatorComponent> ent, ref ReplicatorUpgrade2ActionEvent args)
+    {
+        // don't run this clientside
+        if (_net.IsClient || !_timing.IsFirstTimePredicted)
+            return;
+
+        UpgradeReplicator(ent, 2);
+
+        QueueDel(ent);
+        QueueDel(args.Action);
+    }
+
+    public void OnUpgrade3(Entity<ReplicatorComponent> ent, ref ReplicatorUpgrade3ActionEvent args)
+    {
+        // don't run this clientside
+        if (_net.IsClient || !_timing.IsFirstTimePredicted)
+            return;
+
+        UpgradeReplicator(ent, 3);
+
+        QueueDel(ent);
+        QueueDel(args.Action);
+    }
+
+    public void UpgradeReplicator(Entity<ReplicatorComponent> ent, int desiredLevel)
+    {
+        var xform = Transform(ent);
+
+        // if adding more stages, maybe make this a switch.
```

Comment:
an array or something would probably be better instead of having individual variables for each stage, but given the low count here, it's not a major issue